File: examples/rule110.py
import time


BOARD_SIZE = 100

curBoard = ""

for i in range(BOARD_SIZE - 1):
    curBoard = curBoard + " "

curBoard = curBoard + "x"

newBoard = curBoard

MAX_GENS = 100

def getVal(x1, x2, x3):
    # 7
    if(x1 == "x" and x2 == "x" and x3 == "x"):
        return " "
    # 6
    if(x1 == "x" and x2 == "x" and x3 == " "):
        return "x"
    # 5
    if(x1 == "x" and x2 == " " and x3 == "x"):
        return "x"
    # 4
    if(x1 == "x" and x2 == " " and x3 == " "):
        return " "
    # 3
    if(x1 == " " and x2 == "x" and x3 == "x"):
        return "x"
    # 2
    if(x1 == " " and x2 == "x" and x3 == " "):
        return "x"
    # 1
    if(x1 == " " and x2 == " " and x3 == "x"):
        return "x"
    # 0
    if(x1 == " " and x2 == " " and x3 == " "):
        return " "
    
def advanceGeneration():
    global curBoard
    newBoard = " "
    for i in range(1, BOARD_SIZE - 1):
        newBoard = newBoard + getVal(curBoard[i - 1], curBoard[i], curBoard[i + 1])
    

    newBoard = newBoard + "x"

    curBoard = newBoard


def printCurrentGen():

    print(curBoard)


t1 = time.time()

for i in range(MAX_GENS):
    advanceGeneration()
    printCurrentGen()


t2 = time.time()

print("110 rule took the following seconds")
print(t2 - t1)

File: examples/rule110.cpp
#include <iostream>
#include <cassert>

static constexpr auto BOARD_SIZE = 10;

int currBoard[BOARD_SIZE];
int newBoard[BOARD_SIZE];

int getVal(int x1, int x2, int x3)
{
    // 7
    if(x1 && x2 && x3)
    {
        return 0;
    }

    // 6
    if(x1 && x2 && x3 == 0)
    {
        return 1;
    }

    // 5
    if(x1 && x2 == 0 && x3)
    {
        return 1;
    }

    // 4
    if(x1 && x2 == 0 && x3 == 0)
    {
        return 0;
    }

    // 3
    if(x1 == 0 && x2 && x3)
    {
        return 1;
    }

    // 2
    if(x1 == 0 && x2 && x3 == 0)
    {
        return 1;
    }

    // 1
    if(x1 == 0 && x2 == 0 && x3)
    {
        return 1;
    }

    // 0
    if(x1 == 0 && x2 == 0 && x3 == 0)
    {
        return 0;
    }

    assert(0 && "unreachable");

    return -1;
}

void initBoard()
{
    for(auto i = 0; i < BOARD_SIZE; i++)
    {
        currBoard[i] = 0;
    }
    currBoard[BOARD_SIZE - 1] = 1;
}

void advanceGeneration()
{
    for(auto i = 1; i < BOARD_SIZE - 1; i++)
    {
        newBoard[i] = getVal(currBoard[i - 1], currBoard[i], currBoard[i + 1]);
    }
    newBoard[BOARD_SIZE - 1] = 1;

    for(auto i = 0; i < BOARD_SIZE; i++)
    {
        currBoard[i] = newBoard[i];
    }
}

void printCurrentGen()
{
    for(auto i = 0; i < BOARD_SIZE; i++)
    {
        std::cout << (currBoard[i] ? 'x' : ' ');
    }

    std::cout << std::endl;
}

int main()
{
    static constexpr auto generations = 10;

    initBoard();

    for(auto i = 0; i < generations; i++)
    {
        advanceGeneration();
        printCurrentGen();
    }


    return 0;
}